### PR TITLE
implement writing app state - pin/mute/archive chat

### LINKED ISSUE #31: Implement writing app state
If we write one of these functions, we write the rest. App state write so important.

### PR DESCRIPTION
**Pull request by crazycodezombie**
Fixes #31

In this pull request I added the implementation of writing the app state.

This adds support for:
- Archive chat
- Pin chat
- Mute chat

* other types of states (e.g. delete) can easily be added by adding a PatchInfo definition and few function similar to the pin/mute/archive for the required action.

Archive chat is more complicated than mute/pin and requires a MessageRange parameter which complicates the implementation and need more research for implementing the correct message range. For now I just used the latest message in this parameter and it seems to work, but can be risky in the future.

In addition I added "upgrade5" function to the store, but with the implementation of storing the last message for each chat it will require re-login from the user (and i didn't want to let the user implement the message-range part). not sure what can be a better solution for that.

### CODE REVIEW

File: appstate.go [hunk 1]
```diff
@@ -301,3 +301,79 @@ func (cli *Client) requestAppStateKeys(ctx context.Context, rawKeyIDs [][]byte)
...
+		return err
+	}
+
+	respCollection := resp.GetChildByTag("sync", "collection")
+	respCollectionAttr := respCollection.AttrGetter()
+	if t, ok := respCollectionAttr.GetString("type", false); ok && t == "error" {
+		return fmt.Errorf("unable to update chat, got error response from server")
+	}
+	
+	return cli.FetchAppState(patch.Type, false, false)
```
Comment (crazycodezombie) [inline appstate.go:378]: To me, It seems to be the easiest way to store the new state after writing the state successfully, but when looking on whatsapp-web, it seems to fetch the state only on archive request, I guess on mute/pin they just storing the calculated state from here. anyway fetching the state from the server does the work.

File: message.go [hunk 2]
```diff
@@ -315,6 +315,11 @@ func (cli *Client) handleHistorySyncNotification(notif *waProto.HistorySyncNotif
 			go cli.handleHistoricalPushNames(historySync.GetPushnames())
 		} else if len(historySync.GetConversations()) > 0 {
 			go cli.storeHistoricalMessageSecrets(historySync.GetConversations())
+
+			if historySync.GetSyncType() == waProto.HistorySync_INITIAL_BOOTSTRAP ||
```
Comment (crazycodezombie) [inline message.go:319]: Here I was not sure, HistorySync_INITIAL_BOOTSTRAP & waProto.HistorySync_RECENT(chunk-order-1) is the only cases we should store the last message

**Comment by 1337-sys** [comment]
I tested it works great. @tulir doesn't want to process history messages, so we can give the last message id and archive it, just like in read chat. Please share me your IBAN dude. I waited a year for this moment. You are a perfect <3

File: appstate.go (hunk 1 above)
Comment (tulir) [reply appstate.go:360]: Yeah probably fine to do that for now

File: message.go (hunk 2 above)
Comment (tulir) [reply message.go]: I removed the message storage entirely since it's not necessary

**Review by tulir — APPROVED** [review]

**Comment by Reyuki-san** [comment]
> but with the implementation of storing the last message for each chat it will require re-login from the user

@crazycodezombie i still confused why it will require re-login? could you explain more detail about the reason? i apologize for my poor knowledge

**Pull request merged**